File: src/Python/208.ImplementTrie(PrefixTree).py

class TrieNode:
    def __init__(self):    
        self.__R = 26
        self.__links = [None] * self.__R
        self.__isEnd = False
    
    def containsKey(self, ch):
        return self.__links[ord(ch) - ord('a')] is not None
    

    def get(self, ch):
        return self.__links[ord(ch) - ord('a')]


    def put(self, ch, node):
        self.__links[ord(ch) - ord('a')] = node


    def setEnd(self):
        self.__isEnd = True


    def isEnd(self):
        return self.__isEnd


class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()
        

    def insert(self, word):
        """
        Inserts a word into the trie.
        """
        node = self.root
        for c in word:
            if node.containsKey(c) == False:
                node.put(c, TrieNode())
            node = node.get(c)
        node.setEnd()
        

    def __searchPrefix(self, word):
        node = self.root
        for c in word:
            if node.containsKey(c) == True:
                node = node.get(c)
            else:
                return None
        return node


    def search(self, word):
        """
        Returns if the word is in the trie.
        """
        node = self.__searchPrefix(word)
        return node is not None and node.isEnd()
        

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        node = self.__searchPrefix(prefix)
        return node is not None


if __name__ == "__main__":
    obj = Trie()
    word = 'dugu'
    prefix = 'dugulala'
    obj.insert(word)
    print(obj.search(word))
    print(obj.startsWith(prefix))

File: src/CSharp/208.ImplementTrie(PrefixTree).cs

using System;
using System.Collections.Generic;

namespace PeefyLeetCode.ImplementTrie
{
public class TrieNode {
    private TrieNode[] links;
    private int R = 26;
    private bool _isEnd;

    public TrieNode() {
        links = new TrieNode[R];
    }

    public bool containsKey(char ch) {
        return links[ch -'a'] != null;
    }

    public TrieNode get(char ch) {
        return links[ch -'a'];
    }

    public void put(char ch, TrieNode node) {
        links[ch -'a'] = node;
    }

    public void setEnd() {
        _isEnd = true;
    }

    public bool isEnd() {
        return _isEnd;
    }
}
public class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }
    
    public void Insert(string word) {
        TrieNode node = root;
        for (int i = 0; i < word.Length; i++) {
            char currentChar = word[i];
            if (!node.containsKey(currentChar)) {
                node.put(currentChar, new TrieNode());
            }
            node = node.get(currentChar);
        }
        node.setEnd();
    }
    
    private TrieNode searchPrefix(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.Length; i++) {
           char curLetter = word[i];
           if (node.containsKey(curLetter)) {
               node = node.get(curLetter);
           } else {
               return null;
           }
        }
        return node;
    }

    public bool Search(string word) {
        TrieNode node = searchPrefix(word);
        return node != null && node.isEnd();
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public bool StartsWith(string prefix) {
        TrieNode node = searchPrefix(prefix);
        return node != null;
    }
}

}